activerecord/lib/active_record/relation.rb

```diff
@@ -730,12 +731,14 @@ def values
     end
 
     def inspect
-      subject = loaded? ? records : self
-      entries = subject.take([limit_value, 11].compact.min).map!(&:inspect)
+      @inspect ||= begin
+        subject = loaded? ? records : self
```

Comment:
Isn't this line already doing the memoization? It should only go once to the database.

Reply:
No, if the relation is not yet loaded the subject.take triggers a limit query each time inspect is called.

Reply:
Right, because it's essentially doing `self.take` (or `relation.take`), which does not record it's been already "loaded". What if we forced `load` every time, or simply used `records` directly here as "subject", which loads if not already loaded,  instead of this conditional?

Reply:
Just using normal load behavior would work. The idea of the original code was to load less records for inspect, since only the first 10 are shown in the inspect output. 

It would certainly be simpler, if we removed that logic and always loaded the original relation and do the limiting in ruby.

Reply:
Of course, good point, the original intention _was_ to trigger a SQL _with the limit_ to avoid loading the whole thing unnecessarily. It just happens so that there are places that call inspect multiple times. Sorry I missed that in my previous comment. It does make me wonder if the relation is the place to cache the inspect result, since the callers aren't doing that work of caching it, but I can't think of another answer for that off the top of my head.

Reply:
I've tried to look at the problem from a different perspective:

If I inspect something, I would like to see it's current state, not trigger side-effects.

What do you think about the following logic:

* If the relation is loaded, show the first 10 records in the inspect output.
* If the relation is *not* loaded, show no records (or empty records collection)

Additionally it might be worthwile to include the loaded flag in the inspect output.

This way we never do unneccessary queries and calling inspect on unloaded relations is much cheaper and free of side effects and we can skip the memoization.

Reply:
> * If the relation is _not_ loaded, show no records (or empty records collection)

This is going to degrade the experience we want to give to the Rails console.

`User.all` in a console should records, not show nothing.

I'm not convinced that we need to memoize here.

Unless the `4x with the default middleware stack` are happening in production, or it makes development unbearable, it is simpler for the code and as behavior to not memoize.

Reply:
> This is going to degrade the experience we want to give to the Rails console.
>
> User.all in a console should records, not show nothing.
>
> I'm not convinced that we need to memoize here.

So would you be ok with triggering a load on inspect then, instead of doing a separate limit query?

> Unless the 4x with the default middleware stack are happening in production, or it makes development unbearable, it is simpler for the code and as behavior to not memoize.

On the app I was working on the default middleware stack caused about 200-300 ms additional delay without memoization, which is not so bad, but due to spamming the log with multiple pages of query logs it was very hard to find the actual error output in the logs, since it was before all the inspect calls.

With both meta-request and better_errors active the inspect calls caused several seconds additional delay and easily exhausted the 10,000 line scroll back buffer in my terminal.

Better Errors certainly has its own problems, since it sometimes does 100 inspects per ivar and additional ones when navigating the stack, so I've replaced it with web-console for now.

The instance variables with unloaded relations contain some deeply nested preloading for rendering navigation that is usually only loaded if the relevant fragment caches get busted.

Since the app uses UUID PKs the SQL for preload queries can become quite long.

Reply:
> So would you be ok with triggering a load on inspect then, instead of doing a separate limit query?

I think that was the case before https://github.com/rails/rails/commit/3d6176fded424a2195e399d031e343bbcd0edcad.

We will fix one problem and add another problem.

I'm assuming our default middleware stack needs to inspect those things but I don't remember where those inspect calls happen. Can you point them out?

Reply:
> I'm assuming our default middleware stack needs to inspect those things but I don't remember where those inspect calls happen. Can you point them out?

Thanks for the suggestion. Actually all inspect calls with the default middleware stack are caused by the did_you_mean gem that comes bundled with ruby:

> ActiveRecord::Relation#inspect called 3 times from /Users/felix/.rvm/rubies/ruby-2.6.5/lib/ruby/gems/2.6.0/gems/did_you_mean-1.3.0/lib/did_you_mean/core_ext/name_error.rb:8:in `inspect'
>
> NameError (undefined local variable or method `xxx' for #<ProductFamiliesController:0x00007fd28e689c10>):

Reply:
I spoke too soon, the did_you_mean gem was cloaking the original call locations, with `RUBYOPT="--disable-did_you_mean"` things became clearer:

* 1x inspect from `ActiveSupport::Notifications::Instrumenter#instrument` caused by the line `payload[:exception] = [e.class.name, e.message]` (activesupport-6.0.2.1/lib/active_support/notifications/instrumenter.rb:26:in `inspect')
* 1x inspect from `Airbrake::NestedException#as_json` calling `exception.message` (yeah, not default middleware, forgot to remove that one)
* 1x inspect from `ActionDispatch::DebugExceptions#log_error` caused by `message << "#{exception.class} (#{exception.message}):"` (actionpack-6.0.2.1/lib/action_dispatch/middleware/debug_exceptions.rb:150:in `inspect')

The actual calls are triggered by calling `NameError#message`, so it can't really be avoided and anything working with exceptions will trigger the inspect calls.

Reply:
I've inspected the MRI C code for NameError and what happens is this:

`NameError#message` will call `#inspect` on the object where the name error occurred, unless it is `true`, `false` or `nil`. Since the exception in this case happened in the controller, it will call inspect on the controller, which in turn calls inspect on all its instance variables. Then it checks to see if the result of the inspect call is longer than 65 bytes and if yes, it just calls `#to_s` and uses that in the error message, which means all the work done by inspect was for naught.

Reply:
I've also found out why the problem gets so much worse when using better_errors.

The first problem is that by default Better Errors inspects the `@_request` and `@_response` ivars, which both hold 14 references to the controller instance, so each trigger 14 additional inspect calls per ivar. These can be worked around by setting `BetterErrors.ignored_instance_variables = [:@_request, :@_response]` in an initializer.

The second problem occurs only from the second request after an error, because `BetterErrors::Middleware` stores its error state / session in an `@error_page` ivar on the middleware instance, causing 45 references per ivar when `NameError#message` is called, because it ends up being referenced in the ivars of the controller.

That last problem can easily be fixed by resetting the `@error_page` ivar in the middleware prior to each normal request, so I'm going to submit a PR to better_errors to fix that.